src/v/iceberg/filesystem_catalog.h

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "base/seastarx.h"
+#include "cloud_io/remote.h"
+#include "iceberg/catalog.h"
+#include "iceberg/partition.h"
+#include "iceberg/schema.h"
+#include "iceberg/table_io.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/transaction.h"
+#include "model/fundamental.h"
+#include "utils/named_type.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+namespace iceberg {
+
+// A simple catalog implementation that uses object storage to manage updates
+// to the table metadata. The file layout is similar to the Apache Iceberg
+// HadoopCatalog, which places metadata into per-table directories. This means
+// that metadata written with this catalog can be read by applications that
+// know how to read from the HadoopCatalog.
+//
+// Like the HadoopCatalog, this catalog advertises the latest metadata with a
+// "version hint", a file containing an integer that refers to the latest table
+// metadata. Note, this version is an artifact of the HadoopCatalog only, and
+// is not a part of the table metadata spec.
+//
+// The files laid out by this catalog look roughly like:
+//
+// /warehouse_location/my_namespace/my_table/metadata/version-hint.text
+//   - the version hint: contains the string "42"
+// /warehouse_location/my_namespace/my_table/metadata/v42.metadata.json
+//   - contains the latest table metadata as JSON
+// /warehouse_location/my_namespace/my_table/metadata/v41.metadata.json
+// /warehouse_location/my_namespace/my_table/metadata/v40.metadata.json
+// /warehouse_location/my_namespace/my_table/metadata/v...
+//   - contain older versions of the table metadata
+//   - Iceberg has the write.metadata.delete-after-commit.enabled property to
+//     delete these, though this is not yet implemented by filesystem_catalog
+//
+// Unlike the HadoopCatalog, which uses atomic renames to avoid overwriting
+// existing table metadata, this catalog is by comparison less safe, performing
+// best-effort non-atomic operations to discourage concurrent updates.
+//
+// WARNING: until the concurrency issues are addressed, this catalog should
+// only be used for testing and non-production workloads.
```

Comment:
do you pass in a reference to the catalog implementation as a pointer? if so, this could be a testing-only library?

Reply:
Yeah there will probably be some kind of factory that returns a `std::unique_ptr<catalog>` that is either this or a `catalog` impl that calls out to a REST catalog service.

Maybe I should clarify, this is not just meant for unit tests, but rather for folks who want to try Iceberg out without standing up an external catalog service.